github/oauth_server.go

```diff
@@ -13,82 +17,107 @@ import (
+	githubGrantType     = "urn:ietf:params:oauth:grant-type:device_code"
 )
 
 func NewClientFromToken(token *oauth2.Token) *githubclient.Client {
+	oauthConf := oauth2.Config{
+		ClientID: oauthClientId,
+		Endpoint: githuboauth.Endpoint,
+	}
 	oauthClient := oauthConf.Client(oauth2.NoContext, token)
 	return githubclient.NewClient(oauthClient)
 }
 
 func Authenticate() (*oauth2.Token, error) {
-	ln, err := net.Listen("tcp", ServerBind)
+
+	resp, err := http.Post(fmt.Sprintf("%s?client_id=%s&scope=%s", githubStageOneUri, oauthClientId, oauthScopes), "application/json", nil)
 	if err != nil {
 		return nil, err
 	}
+	defer resp.Body.Close()
 
-	handler := githubAuthHandler{
-		OAuthStateString: "da39a3ee5e6b4b0d3255bfef95601890afd80709",
-		TokenCh:          make(chan *oauth2.Token),
+	if resp.StatusCode != 200 {
+		return nil, errors.New(fmt.Sprintf("Github returned non-200 response (%d)", resp.Status))
 	}
+	body, err := ioutil.ReadAll(resp.Body)
 
-	go func() {
-		_ = http.Serve(ln, handler)
-	}()
+	fmt.Println("Response body:", string(body))
 
-	if err := open.Run(ServerAddress); err != nil {
+	params, err := url.ParseQuery(string(body))
+	if err != nil {
 		return nil, err
 	}
 
-	select {
-	case token := <-handler.TokenCh:
-		return token, ln.Close()
-	case err := <-handler.ErrCh:
+	deviceCode := params["device_code"][0]
+	fmt.Println("Device Code:", deviceCode)
+
+	userCode := params["user_code"][0]
+	fmt.Println("User Code:", userCode)
+
+	verificationUri := params["verification_uri"][0]
+	fmt.Println("Verification URI:", verificationUri)
+
+	expiresIn := params["expires_in"][0]
+	fmt.Println("Expires In:", expiresIn)
+
+	expiresInInt, err := strconv.Atoi(expiresIn)
+	if err != nil {
 		return nil, err
 	}
-}
 
-type githubAuthHandler struct {
-	OAuthConf        *oauth2.Config
-	OAuthStateString string
-	TokenCh          chan *oauth2.Token
-	ErrCh            chan error
-}
+	expiresAt := time.Now().Add(time.Second * time.Duration(expiresInInt))
+	fmt.Println("Expires At:", expiresAt)
+
+	interval := params["interval"][0]
+	fmt.Println("Interval:", interval)
 
-func (handler githubAuthHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	switch req.URL.Path {
-	case `/`:
-		http.Redirect(w, req, "/login", http.StatusMovedPermanently)
-	case `/login`:
-		url := oauthConf.AuthCodeURL(handler.OAuthStateString, oauth2.AccessTypeOnline)
-		http.Redirect(w, req, url, http.StatusTemporaryRedirect)
-	case `/github_oauth_cb`:
-		// check the oauth state string matches, prevent replay attacks
-		if state := req.FormValue("state"); state != handler.OAuthStateString {
-			err := fmt.Errorf("Invalid oauth state, expected %q, got %q", handler.OAuthStateString, state)
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			handler.ErrCh <- err
-			return
+	intervalInt, err := strconv.Atoi(interval)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := open.Run(verificationUri); err != nil {
+		return nil, err
+	}
+
+	for {
```

Comment:
There's probably a fancy Go way using a ticker and a timer, and `select` on the two channels.
But this is also fine :)

Reply:
I dabbled with this for a while, but then realised that github will return an error when the 15 minutes is up and the user token has expired. As a result I've turned this into an endless loop that exits on either:

1. an access token is generated
2. github returns an error, one of which can be "time is up, try again"